graph/nullable.md

```diff
@@ -0,0 +1,312 @@
+Below are some examples of nullable and non-nullable properties. 
+
+## CSDL
+
+```xml
+<EntitySet Name="servicePrincipals" Type="self.servicePrincipal" />
+...
+<EntityType Name="servicePrincipal">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+
+  <Property Name="appId" Type="Edm.String" Nullable="false" /> <!--required for creation-->
+  <Property Name="displayName" Type="Edm.String" Nullable="false" />
+  <Property Name="foo" Type="Edm.String" Nullable="true" DefaultValue="testval" />
+  <Property Name="bar" Type="Edm.String" Nullable="false" DefaultValue="differentvalue" />
+  ...
+</EntityType>
+```
+
+## HTTP Requests
+
+### {1} Create a servicePrincipal with no properties
+
+```HTTP
+POST /servicePrincipals
+
+400 Bad Request
+{
+  "error": {
+    "code": "badRequest",
+    "message": "The 'appId' property is required to create a servicePrincipal."
+  }
+}
+```
+
+### {2} Create a servicePrincipal without a display name
+
+```HTTP
+POST /servicePrincipals
+{
+  "appId": "00000000-0000-0000-0000-000000000001"
+}
+
+201 Created
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "some application name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` was given a value by the service even though no value was provided by the client
+2. `foo` has the default value as specified by its `DefaultValue` attribute in the CSDL
+3. `bar` has the default value as specified by its `DefaultValue` attribute in the CSDL
+
+### {3} Update the display name of a service principal to null
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": null
+}
+
+400 Bad Request
+{
+  "error": {
+    "code": "badRequest",
+    "message": "null is not a valid value for the property 'displayName'; 'displayName' is not a nullable property."
+  }
+}
+```
+Notes:
+1. `displayName` cannot be set to `null` because it has be marked with `Nullable="false"` in the CSDL.
+
+### {4} Update the display name of a service principal
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": "a non-generated display name"
+}
+
+200 OK
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "a non-generated display name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` can be set to any value other than `null`
+2. The response body here is provided for clarity, and is not part of the guidance itself. The [Graph standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) states that the workload can decide the behavior.
```

Comment:
Sorry for being a stickler.  I meant to put a reference to the Graph Guidelines with an idea that the Graph guidelines can be more prescriptive. However, if we use OData as a reference, I think we should display the text as "OData v4.01".

Reply:
Let's use an incremental approach. please change [Graph standard] back to [OData v4.01] and I'll add a backlog item for 'Graph specific HTTP Request/Response guidelines' to the Guidelines project and API Council agenda so that we can close the gap later.

Reply:
I thought that most PATCH requests respond with 204 No content, but using the Prefer representation HTTP request header, the service should respond with a 200 OK and the full response.  Can we just add the Prefer header and the preference applied response header?

Reply:
I changed it back to "OData" for now. Dan, I'm happy to add a prefer header if you think that's best. What I'm trying to convey with this note is that the response body, a prefer header, etc are all just to allow the sample to fully exemplify the behavior; I don't want anyone to be confused that they must have a response body or require a header or anything like that. Let me know, I'll add the headers if you think that's the best way to show that to readers.